### PR TITLE
drivers: mpsl: Add support for MPSL clock driver in glue layer.

### PR DESCRIPTION
**Pull request by sofu1**
Convert the existing glue layer for nrfxlib BLE controller clock driver
into a common glue layer for BLE controller and MPSL clock drivers.
Ifdefs are used to select which API to call from the glue layer.
Add MPSL subsys for proper MPSL initialization, which is required for
LFCLK configuration.
Move multithreading_lock to lib/multithreading_lock to make it
available outside BLE controller.
Modifications to Kconfig and CMakeLists in drivers to include the new
glue layer.

Signed-off-by: Solveig Fure <solveig.fure@nordicsemi.no>

### CODE REVIEW

**Comment by sofu1** [comment]
nrfxlib in manifest file pointing to https://github.com/NordicPlayground/nrfxlib/pull/93

**Comment by sofu1** [comment]
MPSL_relations svg

File: subsys/mpsl/Kconfig [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
```
Comment (SebastianBoe) [inline subsys/mpsl/Kconfig]: This file should be removed.

File: lib/multithreading_lock/Kconfig [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+config MULTITHREADING_LOCK
+	bool "Multithreading Lock"
```
Comment (SebastianBoe) [inline lib/multithreading_lock/Kconfig]: This needs more help text.

File: drivers/clock_control/clock_control_nrfxlib.c [hunk 3]
```diff
@@ -9,8 +9,13 @@
 #include <device.h>
 #include <kernel_includes.h>
 #include <clock_control.h>
+#ifdef CONFIG_BT_LL_NRFXLIB
 #include <ble_controller.h>
 #include <ble_controller_soc.h>
+#else /* CONFIG_MPSL_LIBRARY */
+#include <mpsl.h>
+#include <mpsl_clock.h>
+#endif
```
Comment (SebastianBoe) [inline drivers/clock_control/clock_control_nrfxlib.c]: Eventually, clock_control_nrfxlib will integrate only with MPSL, and not both MPSL and ble_controller, right?

File: subsys/mpsl/mpsl_init.c [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	clock_cfg.source = MPSL_CLOCK_LF_SRC_XTAL;
+#elif CONFIG_CLOCK_CONTROL_NRF_K32SRC_SYNTH
+	clock_cfg.source = MPSL_CLOCK_LF_SRC_SYNTH;
+#else
+#error "Clock source is not supported or not defined"
+#endif
+
+	clock_cfg.rc_ctiv = MPSL_RECOMMENDED_RC_CTIV;
+	clock_cfg.rc_temp_ctiv = MPSL_RECOMMENDED_RC_TEMP_CTIV;
+
```
Comment (joerchan) [inline subsys/mpsl/mpsl_init.c]: clock accuracy is gone.
CLOCK_CONTROL_NRF_ACCURACY Kconfig option does not do anything. Should we remove the option for MPSL?

File: subsys/mpsl/mpsl_init.c [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	err = mpsl_init(&clock_cfg, MPSL_LOW_PRIO_IRQ, m_assert_handler);
+	if (err) {
+		return err;
+	}
+
+	irq_enable(MPSL_LOW_PRIO_IRQ);
```
Comment (joerchan) [inline subsys/mpsl/mpsl_init.c]: Where is the IRQ connected? Documentation for MPSL says that `mpsl_low_priority_process` shall be called when this IRQ is pended.

File: subsys/mpsl/mpsl_init.c [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+#ifdef CONFIG_CLOCK_CONTROL_NRF_K32SRC_RC
+	clock_cfg.source = MPSL_CLOCK_LF_SRC_RC;
+#elif CONFIG_CLOCK_CONTROL_NRF_K32SRC_XTAL
+	clock_cfg.source = MPSL_CLOCK_LF_SRC_XTAL;
+#elif CONFIG_CLOCK_CONTROL_NRF_K32SRC_SYNTH
+	clock_cfg.source = MPSL_CLOCK_LF_SRC_SYNTH;
+#else
+#error "Clock source is not supported or not defined"
+#endif
```
Comment (joerchan) [inline subsys/mpsl/mpsl_init.c]: Extract to function, config_clock_source_get?

File: drivers/clock_control/CMakeLists.txt [hunk 7]
```diff
@@ -5,4 +5,4 @@
 #
 
 zephyr_library()
-zephyr_library_sources(clock_control_ll_nrfxlib.c)
+zephyr_library_sources(clock_control_nrfxlib.c)
```
Comment (joerchan) [inline drivers/clock_control/CMakeLists.txt]: Maybe better to have clock_control_mpsl if the driver is to use this service layer?

File: drivers/clock_control/clock_control_nrfxlib.c (hunk 3 above)
Comment (sofu1) [reply drivers/clock_control/clock_control_nrfxlib.c]: Yes, that's the plan.

File: lib/multithreading_lock/Kconfig (hunk 2 above)
Comment (sofu1) [reply lib/multithreading_lock/Kconfig]: Yes, it should be hidden I think. So I'll remove the prompt. But do you want me to add a help text? Or is it ok without?

Comment (SebastianBoe) [reply lib/multithreading_lock/Kconfig]: A one sentence help text would be good to have.

File: drivers/clock_control/CMakeLists.txt (hunk 7 above)
Comment (sofu1) [reply drivers/clock_control/CMakeLists.txt]: Yes, but it is also for ble_controller right now

File: drivers/clock_control/Kconfig [hunk 8]
```diff
@@ -1,13 +1,15 @@
...
 #
 
-config CLOCK_CONTROL_NRF_LL_NRFXLIB
-	bool "nRF BLE Controller clock control driver"
+config CLOCK_CONTROL_NRFXLIB
+	bool "Enable nrfxlib clock control driver, for use with BLE Controller or MPSL clock control."
 	default y
+	depends on BT_LL_NRFXLIB || MPSL_LIBRARY
 	select CLOCK_CONTROL_NRF_FORCE_ALT
+	select MULTITHREADING_LOCK
```
[the remark this answers is not in the document]
Comment (sofu1) [reply drivers/clock_control/Kconfig]: I added the MULTITHREADING_LOCK select for ble controller in nrfxlib, but I can remove it from there, and add it to the drivers that need it instead.

[the remark this answers is not in the document]
Comment (joerchan) [reply drivers/clock_control/Kconfig]: It seems more correct to me at least to have the select in the driver.

File: drivers/clock_control/CMakeLists.txt (hunk 7 above)
Comment (joerchan) [reply drivers/clock_control/CMakeLists.txt]: Ok. leave as is for now then.

File: subsys/mpsl/mpsl_init.c (hunk 4 above)
Comment (sofu1) [reply subsys/mpsl/mpsl_init.c]: see https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1390/files#diff-2bf5c014915078f1977935e80a88784eR38

File: subsys/mpsl/mpsl_init.c (hunk 5 above)
Comment (sofu1) [reply subsys/mpsl/mpsl_init.c]: I have added it now

**Comment by sofu1** [comment]
Reviewers; please let me know if there is anything else I should do in this PR.

**Comment by SebastianBoe** [comment]
DNM until manifest uses a SHA

File: subsys/mpsl/mpsl_init.c (hunk 5 above)
Comment (joerchan) [reply subsys/mpsl/mpsl_init.c]: IRQ_CONNECT is mapping `SIGNALLING_Handler`, not `mpsl_low_priority_process`. I'm not sure which one is correct. But if we want to have it like in BLE_Controller with the SIGNALLING_Handler we need the handler the semaphore and the signal thread. https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/master/subsys/bluetooth/controller/hci_driver.c#L285

**Pull request merged**